### ISSUE TITLE
Cannot update a nested attribute

### ISSUE THREAD

**Issue by nrf110** [guardian/scanamo#88]
Prior to the 0.9.0 release, we were providing our own UpdateExpression typeclass implementation (NestedSetExpression) to handle this.  With 0.9.x, UpdateExpression was changed from a typeclass-based approach to a sealed trait hierarchy, so we can no longer provide our own implementation.

Is there a way to handle this in 0.9 that I'm missing?  If not, what would be the preferred approach?  Is it possible to have the existing SetExpression handle nested fields, i.e. set("myPage.isViewed", true)?  Or would it be better to have a separate implementation, as we did (i.e. something like setN instead of set)?

**Comment by philwills** [comment]
Interesting, thanks for raising this and sorry for breaking it. 

Whilst I tend to try and avoid overloading, my preference would probably be towards just making it work for the existing methods.

**Comment by philwills** [comment]
OK, this isn't quite as simple as I thought it might be due to the joys of http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/ExpressionPlaceholders.html#ExpressionAttributeNames, but I've got a plan. I'm thinking you'd write something of the form:

```scala
outers.update('id -> id, set('inner \ 'field -> "value"))
```

Does that seem like a sensible approach? I think it would be nice to avoid encoding the separator within the string.

**Comment by nrf110** [comment]
Thanks for the quick response!

I definitely prefer your approach over encoding it in the string, and it covers our existing simple use cases.  To make this handle all nested fields though, I think there is still the case of updating elements within a List, i.e. foo.bar[2] = baz

Extending your example a bit, how do you feel about something like this? (obviously we'd need to add a few more implicit conversions):

```
sealed trait PathNode {
  def name: String
  def separator: String
  def placeholder: String
}

case class KeyPathNode(s: Symbol) extends PathNode{
  def name: String = s.name
  def separator = "."
  def placeholder: String = s"#update${s.name}"
}

case class IndexPathNode(i: Int) extends PathNode {
  def name: String = s"[$i]"
  def separator: String = s""
  def placeholder: String = name
}

case class Field(traversal: NonEmptyVector[PathNode]) {
  val placeholder: String = traversal.foldLeft("")((l, r) => s"$l${r.separator}${r.placeholder}")
  val name: String = traversal.tail.foldLeft(traversal.head.name)((l, r) => s"$l${r.separator}${r.name}")
  val attributeNames: Map[String, String] = { // TODO: Not sure if/how we are supposed to map the list indices
    traversal.filter {
      case _: KeyPathNode => true
      case _ => false
    }.map(node => node.placeholder -> node.name).toMap
  }

  def \ (inner: Symbol): Field = Field(traversal.append(KeyPathNode(inner)))
  def apply(idx: Int): Field = Field(traversal.append(IndexPathNode(idx)))
}
```

usage would then look like:
```
outers.update('id -> id, ('inner \ 'field)(2) -> "value")
```

**Comment by philwills** [comment]
Thanks @nrf110, really appreciate the feedback. I've added a bit to the PR to support updating list elements, which surprised me by leading me to make things simpler.

**Comment by nrf110** [comment]
The implementation looks good, I like the simplification. 👍

**Comment by philwills** [comment]
Closing this, as there's now an alternative available in 0.9.1
